airbyte-cdk/python/airbyte_cdk/sources/declarative/declarative_stream.py

```diff
@@ -172,3 +173,32 @@ def get_cursor(self) -> Optional[Cursor]:
         if self.retriever and isinstance(self.retriever, SimpleRetriever):
             return self.retriever.cursor
         return None
+
+    def _get_checkpoint_reader(
+        self,
+        logger: logging.Logger,
+        cursor_field: Optional[List[str]],
+        sync_mode: SyncMode,
+        stream_state: MutableMapping[str, Any],
+    ) -> CheckpointReader:
+        """
+        This method is overridden to skip classification for incremental streams
+        """
+        mappings_or_slices = self.stream_slices(
+            cursor_field=cursor_field,
+            sync_mode=sync_mode,  # todo: change this interface to no longer rely on sync_mode for behavior
+            stream_state=stream_state,
+        )
+
+        cursor = self.get_cursor()
+        checkpoint_mode = self._checkpoint_mode
+
+        if cursor is not None:
+            self.has_multiple_slices = True
```

Comment:
This seems like a cursor/stream_slicer concept and not a stream one. Would it make sense to move that on the cursor side? I'm asking because I fear that this is not true for cases like [IncrementalSingleSlice](https://github.com/airbytehq/airbyte/blob/86c662bbe6ce9a80f4a5311e3ddb7e22bd4afd5f/airbyte-integrations/connectors/source-monday/source_monday/components.py#L84-L85) i.e. it is not because we have a cursor that we have multiple slices. That being said, I understand that for the source-monday case I just referenced that it wouldn't change anything as `self.supports_incremental == True` so `self.has_multiple_slices` is not considered but to be more precise, the case I fear is that we would have a cursor without cursor field, something like a NoopCursor.

Reply:
I think the reason it's here is because this `has_multiple_slices` field is utilized for both low-code and regular python sources and it was already being stored on the `Stream` class.

Are you proposing that we have a property method on `DeclarativeStream` or something that calls the method/property on DatetimeBasedCursor + PerPartitionRouter + SubstreamPartitionRouter? If thats the case, I don't have strong opinions either but seems fine

Reply:
Thinking about it today, I would assume it would be on the StreamSlicer interface. 

I think what I'm struggling to understand when I see the name `has_multiple_slices` is basically `could_produce_multiple_slices` or something as it is used to:
* Define if a stream is resumable
* Assign a `SubstreamResumableFullRefreshCursor` cursor instead of a `ResumableFullRefreshCursor` in `HttpStream`
It feels like we wouldn't have to consume slices for classification if we would have the `StreamSlicer` explicitly mention is. This means that it pushes the complexity to the StreamSlicer object to know (which is simpler to implement because each stream slicer should know that but is a bit more challenging in terms of migration...) If we put a default value of "false" my understanding is that the worst case is that it won't be flagged as resumable and therefore won't leverage RFR which is not the worst.

Given that we already have a working solution for that, I'm incline to not change that. However, I think this is still an interesting solution to have in the back of our mind in case we face more issues with that. I don't have a example for the `NoopCursor` I was mentioning so I'm fine with ignoring this concern as well.

Reply:
I've updated the method to only bypass classification for the `PerPartitionCursor`, as the issue is specific to streams with parent streams.

The problem occurs because the classification logic, when used with `itertools.tee`, creates a copy of the stream slices. When `stream_slices` is called the second time, the parent records generated during the classification phase are lost. This happens because `itertools.tee` only buffers the results, preventing the `simple_retriever` from observing and updating the cursor again.

Moving this logic to the cursor would require a major CDK release, as all cursors would need to implement this method.

Reply:
Ok, so basically, the state is set before `read` is called but the first slice is not generated again so we are losing the progress on the state.

My question is mostly: can this also be an issue for non-declarative cursors? If so, do we have a plan for these?

Reply:
Yes, this issue can affect any stream where `stream_slices` updates internal values. I've created [an issue](https://github.com/airbytehq/airbyte-internal-issues/issues/10354) for further investigation.

Reply:
thanks for filing the issue @tolik0 . I definitely agree this can affect Python sources. Although I think the likely case of why we don't see happen w/ python sources often if ever is the base implementation of `HttpSubStream` for python sources never really had been coded to use an incremental parent state and it was basically a full refresh every time. 

Using parent incremental state was a low-code feature we never back ported. I pray that we never have to unless we find a good reason to @maxi297 this is also something we'll need to factor into our design of the substream cursor in concurrent CDK